partiql-logical-planner/src/lower.rs

```diff
@@ -737,18 +722,52 @@ impl<'ast> Visitor<'ast> for AstToLogical {
     fn exit_like(&mut self, _like: &'ast Like) {
         let mut env = self.exit_env();
         assert!((2..=3).contains(&env.len()));
-        let escape = if env.len() == 3 {
-            require_str(require_lit(&env.pop().unwrap())).to_string()
+        let escape_ve = if env.len() == 3 {
+            env.pop().unwrap()
         } else {
-            "".to_string()
+            ValueExpr::Lit(Box::new(Value::String(Box::new("".to_string()))))
         };
-        let pattern = require_str(require_lit(&env.pop().unwrap())).to_string();
+        let pattern_ve = env.pop().unwrap();
         let value = Box::new(env.pop().unwrap());
-        let pattern = Pattern::LIKE(LikeMatch { pattern, escape });
-        self.push_vexpr(ValueExpr::PatternMatchExpr(PatternMatchExpr {
-            value,
-            pattern,
-        }));
+
+        match (&pattern_ve, &escape_ve) {
+            (ValueExpr::Lit(pattern_lit), ValueExpr::Lit(escape_lit)) => {
+                match (pattern_lit.as_ref(), escape_lit.as_ref()) {
+                    (Value::String(pattern), Value::String(escape)) => {
+                        let pattern = Pattern::Like(LikeMatch {
+                            pattern: pattern.to_string(),
+                            escape: escape.to_string(),
+                        });
+                        self.push_vexpr(ValueExpr::PatternMatchExpr(PatternMatchExpr {
+                            value,
+                            pattern,
+                        }));
+                    }
+                    _ => {
+                        let pattern_non_string_literals =
+                            Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
+                                pattern: Box::new(pattern_ve),
+                                escape: Box::new(escape_ve),
+                            });
+                        self.push_vexpr(ValueExpr::PatternMatchExpr(PatternMatchExpr {
+                            value,
+                            pattern: pattern_non_string_literals,
+                        }))
+                    }
+                }
+            }
+            _ => {
+                let pattern_non_string_literals =
+                    Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
+                        pattern: Box::new(pattern_ve),
+                        escape: Box::new(escape_ve),
+                    });
+                self.push_vexpr(ValueExpr::PatternMatchExpr(PatternMatchExpr {
+                    value,
+                    pattern: pattern_non_string_literals,
+                }))
+            }
+        }
```

Comment:
Perhaps pull out the `push_vexpr`. I think it makes it a bit clearer what the logic is.

```suggestion
      let pattern = match (&pattern_ve, &escape_ve) {
          (ValueExpr::Lit(pattern_lit), ValueExpr::Lit(escape_lit)) => {
              match (pattern_lit.as_ref(), escape_lit.as_ref()) {
                  (Value::String(pattern), Value::String(escape)) => Pattern::Like(LikeMatch {
                      pattern: pattern.to_string(),
                      escape: escape.to_string(),
                  }),
                  _ => Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
                      pattern: Box::new(pattern_ve),
                      escape: Box::new(escape_ve),
                  }),
              }
          }
          _ => Pattern::LikeNonStringNonLiteral(LikeNonStringNonLiteralMatch {
              pattern: Box::new(pattern_ve),
              escape: Box::new(escape_ve),
          }),
      };

      let pattern = ValueExpr::PatternMatchExpr(PatternMatchExpr { value, pattern });
      self.push_vexpr(pattern);
```

Reply:
Thanks for the suggestion. Moving the `push_vexpr` out of the match expression made things a lot clearer.